When checking out with multiple products, display refund windows separately in checkout

fredrikekelund opened the pull request:
Having a domain and a plan in the cart at the same time would previously cause a conflict with the refund text in the checkout sidebar. The Pro plan is refundable within 14 days, but a domain is only refundable within 4 days. We resolved this by not specifying the exact number of refund days (instead just saying "Money back guarantee").

With this change, we take a different approach – we show multiple rows of refund window info in the checkout summary (one for every product type). 

I have also changed the checkout summary to never show the string `Unlimited customer support via email`. This string is used in other places for marketing purposes, but there is no need to distinguish between "Unlimited" and not in this context.

#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Test the checkout screen for these scenarios:

| Test case (products in cart)                          | Expected output             |Screenshot|
|------------------------------------|-----------------------------|-----------------------------|
| Pro plan, 1 domain | Two refund rows: 14 days and 4 days | ![Pro plan, domain](https://user-images.githubusercontent.com/1101677/167816357-1074eeb9-d8ed-4b49-b76d-e324f69750f9.png) |
| Pro plan | One refund row: 14 days | ![Pro plan](https://user-images.githubusercontent.com/1101677/167816445-ce81eed4-3bdc-41cf-8a94-e12ce2ea642f.png) |
| 3 domains (different TLDs), Google Workspace monthly plan | Four refund rows: 3 × 4 days, 7 days | ![With pro plan, 3 domains, Google Workspace](https://user-images.githubusercontent.com/1101677/167816948-ccf21c2c-46d8-479e-b94f-57cc25ffaef1.png) |
| 3 domains (different TLDs) | Three refund rows: 3 × 4 days | ![With pro plan, 3 domains](https://user-images.githubusercontent.com/1101677/167817121-b7dbcee0-ae25-4ece-9617-d506c9ec1649.png) |
| 3 domains (same TLD) | One refund row: 4 days | ![3 domains, same TLD](https://user-images.githubusercontent.com/1101677/167820265-110af825-abde-4d10-b8e3-7350099704e9.png) |

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Follow up of #54210

fredrikekelund committed:
Display refund windows individually in checkout

Having a domain and a plan in the cart at the same time would previously cause a conflict. The Pro plan is refundable within 14 days, but a domain is only refundable within 4 days. We resolved this by not specifying the number of refund days (instead just saying "Money back guarantee").

With this change, we display multiple rows of refund window info in the checkout summary – one for every product type.

I have also changed the checkout summary to never show the string "Unlimited customer support via email". This string is used in other places for marketing purposes, but there is no need to distinguish between "Unlimited" and not in this context.

olaseni commented on client/my-sites/checkout/composite-checkout/components/wp-checkout-order-summary.tsx:
```diff
@@ -159,20 +155,25 @@ function CheckoutSummaryFeaturesList( props: {
-	}
-	const refundText = getRefundText( refundDays, null, translate );
+	const uniqueRefundableProducts = refundableProducts.reduce< ResponseCartProduct[] >(
+		( acc, p ) => {
```
You shouldn't use abbreviations or codes for variable names. This will make your code less readable. Use fully spelt out meaningful names that helps to document your code.

fredrikekelund replied:
I was intentionally succinct here to avoid unnecessary line wrapping, but I'll revert this in order to stick to our common practices 👍 And in any case, maybe `acc` is quite cryptic when compared to something like `result`.

olaseni commented on client/my-sites/checkout/composite-checkout/components/wp-checkout-order-summary.tsx:
```diff
@@ -188,37 +189,22 @@ function CheckoutSummaryFeaturesList( props: {
-	const translate = useTranslate();
-
-	if ( hasPlanInCart && ! isJetpackNotAtomic ) {
-		return <span>{ translate( 'Unlimited customer support via email' ) }</span>;
```
This copy may be product specific. I would be more confident if we could ensure that this is not the case. @sirbrillig do you have an idea if we could remove this?

sirbrillig replied:
Hm... the history of that block is quite old and has been updated several times. I don't have a particular opinion here, but you might consider checking with the Jetpack folks. I guess make sure that this issue is still fixed: https://github.com/Automattic/wp-calypso/issues/44793

fredrikekelund replied:
This was discussed briefly in https://github.com/Automattic/wp-calypso/pull/54210#issuecomment-1112293740 

It looks to me like the history behind the `! isJetpackNotAtomic` logic has to do with the fact that Jetpack sites would not receive live chat support and this string was previously `Email and live chat support`. See #44872 and #50943 for some history.

Regarding live chat, the `CheckoutSummaryChatIfAvailable` (in this same file) seems to have taken over the responsibility for determining whether we offer live chat support or not for the products in the cart.

At this point, there doesn't seem to be any meaningful distinction between `Unlimited customer support via email` and `Customer support via email` from the customer's perspective.

olaseni commented on client/my-sites/checkout/composite-checkout/components/wp-checkout-order-summary.tsx:
```diff
@@ -159,20 +155,25 @@ function CheckoutSummaryFeaturesList( props: {
-	const translate = useTranslate();
-	const isJetpackNotAtomic = useSelector( ( state ) =>
-		siteId ? isJetpackSite( state, siteId ) && ! isAtomicSite( state, siteId ) : undefined
+	const refundableProducts = responseCart.products.filter(
```
Displaying multiple rows for domain names with different TLDs looks like duplication. Could we group them, except for the case when the refund window varies.

fredrikekelund replied:
I see what you mean here. It does get a bit repetitive when there are multiple domains in the cart. 

If we want to do this, the main challenge as I see it has to do with product names and translations. The `product_name` property of domains varies depending on TLD, so simply grouping all domains might produce funny results. For example, if I have a `.ninja` and `.dev` in my cart, the order summary might say `4-day money back guarantee for .ninja Domain Registration`. This could get confounding for the customer ("isn't there a money back guarantee for my `.dev` domain?") 

One solution to this problem could be to always pass the string `Domain Registration` for domain products, instead of relying on `product.product_name`. I'm a little unsure about if this will reliably result in a translated product name, though... I will investigate!

In the meantime, I have pushed a fix that modifies the grouping logic.

stephanethomas replied:
While grouping seems like a good idea, I think it's ok to keep them separate if it makes our logic too complex or fragile. We can have a look to data but I bet most users only purchase one domain at a time, so it probably doesn't worth the effort.

fredrikekelund replied:
Ended up rolling back the grouping logic. See my latest comment for a little bit more detail https://github.com/Automattic/wp-calypso/pull/63489#issuecomment-1131869773

olaseni reviewed (changes requested):
I left a couple of suggestions for your consideration.

fredrikekelund committed:
Group products by utility functions instead of product_name

This mainly has the advantage of ensuring that all domains are grouped in the checkout order summary (as opposed to being grouped by TLD)

fredrikekelund committed:
Translate "Domain Registration" product name

daledupreez commented on client/my-sites/checkout/composite-checkout/components/wp-checkout-order-summary.tsx:
```diff
@@ -188,37 +198,28 @@ function CheckoutSummaryFeaturesList( props: {
-			) }
+			{ uniqueRefundableProducts.length &&
+				uniqueRefundableProducts.map( ( product ) => {
+					const productName = isDomainProduct( product )
+						? translate( 'Domain Registration', { textOnly: true } )
+						: product.product_name;
```
Why not use the domain name in `product.meta` to make it clearer what the refund window is in each case? I feel that would be more useful, especially for cases where users may not be 100% clear on what a domain registration is.

fredrikekelund replied:
As it currently stands, if the customer has more than one domain in their cart, we still just display one row in the checkout summary.

If we want to revert to the previous logic however, then I agree that this is probably a better idea.

Any thoughts on collapsing multiple domains into a single row vs showing on row per domain?

daledupreez replied:
My thought was driven by the relevance of the wording to an end user for two reasons:
1. I don't know how often users actually end up buying multiple domains -- most of our users only have one domain total, so are very unlikely to purchase two at the same time
2. When users are buying a domain-related product, it may be more relevant to refer to the domain name itself rather than our internal name for the product, as that may make more immediate sense to the user

Maybe @Automattic/nomado can weigh in with their thoughts on which approach would be better.

fredrikekelund replied:
Ended up rolling back the grouping logic. I think users can tolerate some repetition as long as the relevance to them is clear. By printing the actual domain on each row (instead of `Domain Registration`), I think we can consider the information relevant rather than just repetitive.

See my latest comment for current status https://github.com/Automattic/wp-calypso/pull/63489#issuecomment-1131869773

stephanethomas commented on client/my-sites/checkout/composite-checkout/components/wp-checkout-order-summary.tsx:
```diff
@@ -159,20 +155,34 @@ function CheckoutSummaryFeaturesList( props: {
+			return result;
+		},
+		[]
+	);
```
Maybe we could extract this piece of code (from `refundablePredicates` to `uniqueRefundableProducts`) to a separate function (or hook) in this file? That way it would reduce the length of `CheckoutSummaryFeaturesList`.

stephanethomas reviewed (changes requested):
The code looks good but I'm worried that we're aggregating refund windows per product whereas we should also probably take into account the billing period (yearly or monthly). For example, during my tests I added an additional mailbox for Professional Email yearly and another one for Professional Email monthly but the checkout page would show a 14-day refund window for both (it should be 7 days for monthly):

![image](https://user-images.githubusercontent.com/594356/168650348-cc18cf3b-8599-4956-8786-9fd265d2d50b.png)

As a side note, I'm also wondering if we actually offer refunds for additional mailboxes (i.e. any mailbox purchased after the initial purchase). The [support page](https://wordpress.com/support/refunds/) doesn't mention anything, so I've asked the Email Guild for confirmation in Slack (see p1652722005204539-slack-CJPRDN47M).

I suspect there is a balance to find between showing the exact refund window for each product in the cart, and aggregating this information when it makes sense. I've mentioned [yesterday](https://github.com/Automattic/wp-calypso/pull/63489#discussion_r873163099) that I was worried of the complexity that the later approach may introduce but thinking about it more I'm even more worried of displaying invalid information to users. Ideally we should show the same information than in the [support page](https://wordpress.com/support/refunds/).

@fredrikekelund: I'm not 100% sure what approach would be best here, so I'm happy to discuss those changes in a call when you come back.

fredrikekelund committed:
Skip deduplication of refundable products in checkout order summary

Instead we opt to display more descriptive product names for every CheckoutSummaryFeaturesListItem.

fredrikekelund committed:
When multiple emails in one product, list them all

fredrikekelund commented:
As per the last comment from @stephanethomas, I think there's a pretty strong case for skipping the deduplication entirely. Most users are likely to only purchase one domain at a time and we definitely get simpler logic by not trying to deduplicate the refundable products. I've updated the code accordingly.

So right now, if I have two domains and two yearly professional emails in my cart, I get three `CheckoutSummaryFeaturesListItem`s with refund details in the checkout summary (professional emails appear as one product that contains multiple emails). To avoid repetitive text, we spell out which domain/email address we are talking about on each row.

![lorem](https://user-images.githubusercontent.com/1101677/169335927-2d2c1c69-a3ae-403c-a73e-dc25090dca68.png)

olaseni reviewed (approved):
Based on review considerations - looks good!

michaeldcain commented:
@fredrikekelund: thanks for fixing this and pinging me about the multiple domain lines in the sidebar. 

We spoke in Slack, but for the sake of having it be more public, I wonder if (with this PR and #54210) we're conflating the more marketing-based purpose of the sidebar summary area with the information-based purpose of the (i) blocks at the bottom of Checkout. It's not really documented, but I believe the original purpose of the sidebar area was to show the user the value included with their purchase (in hopes that they'll complete Checkout), whereas the (i) info area serves the purpose of explaining how the subscription works.

With these two PRs, and a common cart consisting of plan + domain + email, a user would see three different lines in the sidebar area, as opposed to the more generic "money back guarantee". If the goal is to help the user understand the terms of the money back guarantee, we could use the more generic refund line and relatively link to the (i) area, using the existing domain refund block (or a new refund block for all products) to explain the subtleties of the policy:

![Screen Shot 2022-06-02 at 9 54 23 AM](https://user-images.githubusercontent.com/942359/171656482-cf3f9b8d-9f6f-4637-840f-1fa0575f5b41.png)
